ipq40xx: mikrotik: introduce dynamic BDF loading

robimarko opened the pull request:
Currently, MikroTik devices under ipq40xx use packaged BDF-s in the API 2 aka board-2.bin format.
However, we have received reports from people complaining that the wireless performance of their hAP ac2 board was much worse than compared to the ROS.

Upon further inspection, it appears that MikroTik started doing what we suspected might happen in the future and that is shipping
updated boards, but without changing the revision and with updated BDF-s to match it.

However, this presents an issue as there is no way for the user to identify which BDF-s are needed and no way to update them easily once running.

So since the MikroTik platform driver already supports fetching the BDF data we can use that to provide the BDF-s on the fly.
However, the issue is that currently there is no way to differentiate the hotplug calls for the BFD-s as the ath10k driver just requests the board.bin to be loaded, and thus the same BDF will be loaded for both radios instead of the per radio BDF-s.

So, to resolve this we extend the ath10k driver to first try getting the BDF with a filename specific to the radio like it does for pre-cal which is named pre-cal-bus-device.bin for example: pre-cal-ahb-a800000.wifi.bin
It will now try fetching BDF with the name formatted as board-bus-device.bin for example: board-ahb-a800000.wifi.bin

chunkeey commented:
Heh, one nuance that gets overlooked all the time is that Qualcomm does this board-2.bin integration for their QCA9377/QCA6174 cards. If Mikrotik had to deal with Qualcomm and upload their board-2.bins I'm sure they would have made this a REV2 variant.

Anyway, back in the days I made the first "board-2.bin" with cat + printf in the shell.

it looked like this (I threw this together just now, but it should produce something very close to a working board-2.bin file):
````
#!/bin/sh

add_entry()
{
	local id="${1:0:2}"
	local file="$2"

	printf "\x00\x00\x00\x00\x58\x2f\x00\x00\x00\x00\x00\x00\x25\x00\x00\x00"
	printf "bus=ahb,bmi-chip-id=0,bmi-board-id=${id}\x00\x00\x00"
	printf "\x01\x00\x00\x00\x20\x2f\x00\x00"
	cat "$file" 
}

(
	printf "QCA-ATH10K-BOARD\x00\x00\x00\x00" 
	add_entry 16 board-1-16.bin
	add_entry 17 board-1-17.bin
) > board-2.bin
````
The reason why this worked was that the length of the IDs are always 2 (for QCA4019 at least)
and the board-files sizes were always 12064 (0x2f20). So the boilerplate content of the printf is
always static.

robimarko commented:
> Heh, one nuance that gets overlooked all the time is that Qualcomm does this board-2.bin integration for their QCA9377/QCA6174 cards. If Mikrotik had to deal with Qualcomm and upload their board-2.bins I'm sure they would have made this a REV2 variant.

Well, those are the consumer PCI variants and they are common in notebooks, etc so on those, they have to make sure that BDF-s are upstream.
On others they couldn't care less, that is why they simply let the ODM-s recycle the same 2 BMI ID-s all over.
They even stopped merging out patches to add more boards upstream a long time ago.

I thought of making board-2.bins on the fly, but I don't like the complexity that it would add especially since ath10k and even ath11k support just plain BDF-s in the board.bin just fine, all that was missing a way to differentiate the sysfs/hotplug request.

> The reason why this worked was that the length of the IDs are always 2 (for QCA4019 at least)
> and the board-files sizes were always 12064 (0x2f20). So the boilerplate content of the printf is
> always static.
This is surprisingly true for almost all ath10k cards as far as I know, it was ath11k that uses 3 for length and much larger BDF-s(128k)

robimarko commented:
> How do we know if the correct board data is loaded?
There is no bulletproof way as MikroTik can always completely change their wlan_data format, but so far we know the offsets at which we can simply read the BDF-s from using the MikroTik platform driver.
I tested this on an hAP ac3 and I always get the same CRC32 as I will get by using the traditional packaged BDF-s that will get shipped as a part of the board-2.bin

@notr1ch Made the report about bad WLAN performance on the hAP ac2 and tested this to confirm that the performance was back to ROS levels.
I could also see that the BDF CRC-s were completely different than the ones we package for hAP ac2 so it makes sense as well.

This obviously needs more testing from as wide an audience as possible, ideally from those having WLAN performance issues.

As far as the default board-2.bin goes, well you identified the issue and that is that the ath10k-ct firmware depends on the board package, and I don't have a clever idea how to not make it hard depend so we can intentionally remove it from MikroTik subtarget but also keep it as a default for everything else so nothing breaks.

john-tho commented:
Removing the board file as part of the pre-cal works for me on hap ac2. I don't know if this might race loading the board file.

```diff
--- a/target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
+++ b/target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
@@ -6,6 +6,17 @@
 
 board=$(board_name)
 
+# force mikrotik devices to use per-device board data
+rm_board_bin() {
+       local brdfile=/lib/firmware/ath10k/QCA4019/hw1.0/board-2.bin
+       # if no ipq-wifi package owns the overlay board-2.bin file
+       if [ -f "$brdfile" -a ! -n "$(opkg search "/rootfs-overlay$brdfile")" ]; then
+               # remove the linux-firmware board-2.bin file
+               # so that the board data must be loaded from hard_config
+               rm "$brdfile"
+       fi
+}
+
 case "$FIRMWARE" in
 "ath10k/cal-pci-0000:01:00.0.bin")
        case "$board" in
@@ -141,6 +152,7 @@ case "$FIRMWARE" in
                ath10k_patch_mac $(macaddr_add $(get_mac_binary "/sys/bus/i2c/devices/0-0050/eeprom" 0x66) 2)
                ;;
        mikrotik,hap-ac2)
+               rm_board_bin
                wlan_data="/sys/firmware/mikrotik/hard_config/wlan_data"
                ( [ -f "$wlan_data" ] && caldata_sysfsload_from_file "$wlan_data" 0x0 0x2f20 ) || \
                ( [ -d "$wlan_data" ] && caldata_sysfsload_from_file "$wlan_data/data_0" 0x0 0x2f20 )
@@ -255,6 +267,7 @@ case "$FIRMWARE" in
                ;;
        mikrotik,hap-ac2|\
        mikrotik,sxtsq-5-ac)
+               rm_board_bin
                wlan_data="/sys/firmware/mikrotik/hard_config/wlan_data"
                ( [ -f "$wlan_data" ] && caldata_sysfsload_from_file "$wlan_data" 0x8000 0x2f20 ) || \
                ( [ -d "$wlan_data" ] && caldata_sysfsload_from_file "$wlan_data/data_2" 0x0 0x2f20 )
```

robimarko commented:
> Should have worded this better. More how do I know which board data is loaded.
> Guessing I should be able to tell from api version in dmesg: ath10k_ahb a000000.wifi: board_file api 2 bmi_id 0:16 crc32 e2dfaa91.
> I wanted to check that the checksums shown by ath10k-ct in cat /sys/kernel/debug/ieee80211/phy0/ath10k/fw_checksums matched the file inputs, but I cannot seem to calculate a match for the CRC32 checksum for any of them using rhash on my computer, with the board-2 or firmware-5 files copied from the device, and eventually check dd if=/sys/firmware/mikrotik/hard_config/wlan_data/data_2 of=/tmp/brd_2.bin iflag=skip_bytes,fullblock bs=$((0x2f20)) skip=$((0x2f20)) count=1

ath10k doesn't really tell you, ath10k-ct copies the BDF name and its in its debugfs: `/sys/kernel/debug/ieee80211/phy0/ath10k/firmware_info`

But you can obviously go off the API version as 1 is plain BDF and 2 is the structured format for multiple BDF-s and matching.

As far as removing the board-2.bin goes, can we maybe simply make variants of the ath10k-ct FW package?
Like one with the BDF and the other without it and just include the one without it in the MikroTik sub-target, that sounds cleaner than removing the file.

f00b4r0 commented:
Removing the board-2.bin is likely racy and wastes space in the image. @robimarko's suggestion sounds good

chunkeey commented:
@f00b4r0 the board-2.bin are provided in separate packages nowadays: ath10k-board-qca4019. If that package is removable then there will be no racy conflicts. Also, from what I can tell, the full board data files are extracted and stored in /lib/firmware was well. So they too need space.

(the raw board-1 stuff was considered "legacy" by QCA back in 2018 already. The whole ipq-wifi package was born out of the necessity because back then they really wanted to continue with board-2. format. So, Let's see if they came to their senses or not. Then again, there's an alternative: just generate the board-2.bin on-the-fly. this already works right now, can be done "per-device" and doesn't need any patches to either ath10k or ath10k-ct.)

chunkeey commented:
in that case, wouldn't it make sense to extend ath10k and the mikrotik-code that exports them via sysfs to get these internally via nvmem? Because then @f00b4r0 wish to not have anything on permanent storage would be possible? I could see that a ath10k-nvmem (which would include board- and precal-) would be very useful for much more devices.

chunkeey commented:
> I'm afraid you don't understand how this whole mikrotik thing works. Very quickly, I wrote a driver that expose the caldata+BDF (which are encoded/compressed on flash) to a sysfs folder, where they are then dynamically loaded to the driver (btw, I wish all devices that can do this would do this instead of wasting precious flash space). There is already nothing on permanent storage.
>
> NVMEM is not an option here (I looked into it when I wrote the driver), the data needs quite some work to be accessed and presented.

I see it now thanks. However my immediate question here is: why are you doing this uncompressing buisness in the kernel at all?
In my opinion, this squarly belongs into userspace, in the same way for example how the AVM FritzBox gear has to do the cal-data extraction or bit-reversal in userspace.

john-tho commented:
Test results: Have been running for four hours with no issues on:
- A hapac2 (a busy AP with many stations), with a (new) wlan_data folder.
  The 5GHz board data crc32 (2c0bb1b6) is different to what is packaged in OpenWrt for it.
- An old sxtsq5ac (point to point), with (old) wlan_data
  The 5GHz board data crc32 (3514aeae) is the same as OpenWrt ships for it.

The build I installed which was based on this PR + Makefile exclude `ath10k-board-qca4019` [from above](https://github.com/openwrt/openwrt/pull/4679#issuecomment-944170532), and rebased on dd0ad9b661


regarding ath10k board crc32:
dmesg for ath10k loading board-2.bin shows the crc32 of the whole board-2.bin file.
The ath10k crc32 values can be calculated using [`ath10k-bdencoder --info`](https://github.com/qca/qca-swiss-army-knife/blob/master/tools/scripts/ath10k/ath10k-bdencoder), or the process from it for api1 / board.bin.
<details>
<summary>python ath10k crc32s</summary>

```python
import hashlib
import binascii

# to workaround annoying python feature of returning negative hex values
def hex32(val):
    return val & 0xffffffff
# match with kernel crc32_le(0, buf, buf_len) implementation
def _crc32(buf):
    return hex32(~(hex32(binascii.crc32(buf, 0xffffffff))))

print('{:08x}'.format(_crc32(board_bin_data_file_contents)))
```
</details>

Would be great to settle on a method of excluding `ath10k-board-qca4019`, and include it in the PR.
Then, are we planning to wait for linux-wireless / ath10k-ct upstream feedback or not?

Cheers

wwortel commented:
@robimarko 
introduced the 4 'dynamic bdf' patches in my repository and compiled for hAPac2. How can I confirm that the calibration data used actually come from /sys/firmware/mikrotik/hard_config/wlan_data/data_0 and data_2 ? In /lib/firmware/ath10k/QCA4019/hw1.0 are the board and firmware files from linux-firmware-20210511.tar.xz

notr1ch commented:
@wwortel You can check the CRC of the board file, with this patch you will see different CRCs for the boards as a consequence of them being loaded per radio. Example:


```
[  668.945300] kmodloader: loading kernel modules from /etc/modules.d/*
[  669.788518] ath10k_ahb a000000.wifi: qca4019 hw1.0 target 0x01000000 chip_id 0x003b00ff sub 0000:0000
[  669.788621] ath10k_ahb a000000.wifi: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[  669.802215] ath10k_ahb a000000.wifi: firmware ver 10.4b-ct-4019-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 6b2b5c5b
[  669.852558] ath10k_ahb a000000.wifi: Loading BDF type 0
[  670.057990] ath10k_ahb a000000.wifi: board_file api 1 bmi_id 0:16 crc32 0c06cf77
                                                                     ^^^^^^^^^^^^^^
[  671.346406] ath10k_ahb a000000.wifi: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
[  671.346488] ath10k_ahb a000000.wifi: msdu-desc: 2500  skid: 32
[  671.382952] ath10k_ahb a000000.wifi: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
[  671.384312] ath10k_ahb a000000.wifi: wmi print 'free: 53252 iram: 13432 sram: 35752'
[  671.541905] ath10k_ahb a000000.wifi: htt-ver 2.2 wmi-op 6 htt-op 4 cal pre-cal-file max-sta 32 raw 0 hwcrypto 1
[  671.676864] ath: EEPROM regdomain sanitized
[  671.676966] ath: EEPROM regdomain: 0x64
[  671.677020] ath: EEPROM indicates we should expect a direct regpair map
[  671.677158] ath: Country alpha2 being used: 00
[  671.677214] ath: Regpair used: 0x64
[  672.503992] ath10k_ahb a800000.wifi: qca4019 hw1.0 target 0x01000000 chip_id 0x003b00ff sub 0000:0000
[  672.504056] ath10k_ahb a800000.wifi: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[  672.517595] ath10k_ahb a800000.wifi: firmware ver 10.4b-ct-4019-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 6b2b5c5b
[  672.567914] ath10k_ahb a800000.wifi: Loading BDF type 0
[  672.778179] ath10k_ahb a800000.wifi: board_file api 1 bmi_id 0:17 crc32 2c0bb1b6
                                                                     ^^^^^^^^^^^^^^
[  674.066656] ath10k_ahb a800000.wifi: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
[  674.066738] ath10k_ahb a800000.wifi: msdu-desc: 2500  skid: 32
[  674.102631] ath10k_ahb a800000.wifi: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
[  674.103935] ath10k_ahb a800000.wifi: wmi print 'free: 53252 iram: 13432 sram: 35752'
[  674.230668] ath10k_ahb a800000.wifi: htt-ver 2.2 wmi-op 6 htt-op 4 cal pre-cal-file max-sta 32 raw 0 hwcrypto 1
[  674.341698] ath: EEPROM regdomain sanitized
[  674.341801] ath: EEPROM regdomain: 0x64
[  674.341854] ath: EEPROM indicates we should expect a direct regpair map
[  674.341992] ath: Country alpha2 being used: 00
[  674.342045] ath: Regpair used: 0x64
[  674.355655] kmodloader: done loading kernel modules from /etc/modules.d/*
```

This PR patches the driver to include the radio PCI address(?) as part of the filename: https://github.com/openwrt/openwrt/pull/4679/files#diff-1bddc00ff662ded68d7bd63aded24b3b1c6bd244482c70721a7a3abd654bfcacR47-R48

As the board file with this new filename does not exist in /lib/firmware for this new filename, they are then fetched dynamically from sysfs:
https://github.com/openwrt/openwrt/pull/4679/files#diff-791d1298683afd04a9672e362da47fa041b090f0dfb13ff438e15e43abb9939dR278-R296

wwortel commented:
thanks for your comments.
Had expected that this board file that gets created by 11-ath10k-caldata would be permanent somewhere but I do not find it. The dynamic process does work but is repeated every boot. Had understood, but perhaps wrongly, that only on first boot this would happen and then later ath10/ath10k-ct would find the file in the /lib/firmware area where the firmware-5.bin is.

f00b4r0 commented:
@wwortel it happens at every boot by design. No space is wasted this way.

wwortel commented:
> @wwortel it happens at every boot by design. No space is wasted this way.
Tnx.
Added at start of '11-ath10k-caldata' a line that stores the $FIRMWARE in a temporary file. Noted that all calls occur twice and there to be 14 different ones, so 28 in total. In combination with caldata's selection on basis of board type this produces 4 pre-cal actions and 4  board-ahb-... actions.
Have not found the origin of the duplication of all these attempts that happen in pairs per call but some time is lost during boot it seems.

f00b4r0 commented:
@wwortel these calls are kernel-generated. Nothing to worry about.

wwortel commented:
> @wwortel these calls are kernel-generated. Nothing to worry about.
Tnx
In that case the only comment I have to make is that one needs to make sure this ath10k-board-qca4019 package does not get compiled in as it interferes with the dynamic process. Will test this patch on other devices like the 'LHG (XL) ac' and 'LDF as' I adapted the patch to include those.

wwortel commented:
Have applied patches to Mikrotik 'LHG 5 ac', 'LHG XL 5 ac', and 'LDF 5 ac', using standard ath10k driver and firmware.
Dynamic loading works fine. Started from current openwrt/openwrt:master and in addition to the 4 patches from this thread needed to unselect ath10k-board-qca4019 via menuconfig. This ath10k-board-qca4019 is currently a default addition to all ipq4019.

robimarko commented:
So finally had some time to rebase this onto master and utilize the BDF package disentanglement.
It should now just work without having to deselect anything, unfortunately, no comments on the ath10k patch so far.
No idea how much it makes sense to wait more.

f00b4r0 commented:
@robimarko I'd remove the "draft" status of this PR if you think it's ready for merge, it won't get any traction otherwise. Whether or not to wait for upstream is a decision to be made by maintainers.

robimarko commented:
@f00b4r0 Yeah, I agree.
Updated the PR message and changed the status

notr1ch commented:
I've been running a build for the last 30 days that includes the dynamic BDF loading on a hAP AC2 and the wireless performance went from unusable to pretty much perfect, so this is a somewhat critical PR for proper wireless support on at least the hAP AC2. I suspect the same issue will eventually affect other ipq40xx based Mikrotik devices such as the hAP AC3, cAP AC and wAP AC (if not already).

Happy to see this PR ready for review and look forward to seeing it in a release. Thanks again @robimarko!

robimarko committed:
ipq-wifi: remove packaged BDF-s for MikroTik devices

Since we now provide the BDF-s for MikroTik IPQ40xx devices on the fly,
there is noneed to include package and ship them like we do now.

This also resolves the performance issues that happen as MikroTik
changes the boards and ships them under the same revision but they
actually ship with and require a different BDF.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
ipq40xx: mikrotik: provide BDF-s on demand

Since we now can pass the API 1 BDF-s aka board.bin to the ath10k
driver per radio lets use that to provide the BDF-s for MikroTik devices.

This also resolves the performance issues that happen as MikroTik changes
the boards and ships them under the same revision but they actually ship
with and require a different BDF.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
mac80211: ath10k: backport bus and device specific API 1 BDF selection

Some ath10k IPQ40xx devices like the MikroTik hAP ac2 and ac3 require the
BDF-s to be extracted from the device storage instead of shipping packaged
API 2 BDF-s.

This is required as MikroTik has started shipping boards that require BDF-s
to be updated, as otherwise their WLAN performance really suffers.
This is however impossible as the devices that require this are release under
the same revision and its not possible to differentiate them from devices
using the older BDF-s.

In OpenWrt we are extracting the calibration data during runtime and we are
able to extract the BDF-s in the same manner, however we cannot package the
BDF-s to API 2 format on the fly and can only use API 1 to provide BDF-s on
the fly.
This is an issue as the ath10k driver explicitly looks only for the board.bin
file and not for something like board-bus-device.bin like it does for pre-cal
data.
Due to this we have no way of providing correct BDF-s on the fly, so lets
extend the ath10k driver to first look for BDF-s in the board-bus-device.bin
format, for example: board-ahb-a800000.wifi.bin
If that fails, look for the default board file name as defined previously.

So, backport the upstream ath10k patch.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
Rebased this on top of master, added hap ac3 conversion as well.
@hauke Is this something you may be interested in?

This is only gonna be affecting more and more devices as there are multiple pending MikroTik IPQ40xx devices.

f00b4r0 reviewed (approved):
LGTM

alaraun reviewed (approved):
LGTM. It just works. My cAP ac have been running with it for weeks.

wwortel reviewed (approved):
tested and deployed on Mikrotik hAP2, hAP3, SXTsq 5ac, LDF 5ac, and LHG 5ac.
No issues found.

alaraun commented:
> Rebased this on top of master, added hap ac3 conversion as well. @hauke Is this something you may be interested in?
> 
> This is only gonna be affecting more and more devices as there are multiple pending MikroTik IPQ40xx devices.

It would be really nice to get this implemented.

alaraun commented:
@john-tho @chunkeey @bdobe please add your approval for this PR then it might get merged :)

chunkeey commented:
@alaraun 

see https://github.com/openwrt/openwrt/pull/4679#issuecomment-943362125
and https://github.com/openwrt/openwrt/pull/4679#issuecomment-942690074

f00b4r0 commented:
> @alaraun 
> 
> 
> 
> see https://github.com/openwrt/openwrt/pull/4679#issuecomment-943362125
> 
> and https://github.com/openwrt/openwrt/pull/4679#issuecomment-942690074

Your point being?

f00b4r0 commented:
@chunkeey 

> oh no :(. From what I can tell, there's an issue here with that "policy vs mechanism".
> 
> Your https://patchwork.ozlabs.org/project/openwrt/cover/20200420133503.18700-1-hacks@slashdirt.org/ includes this extract:
> 
> > On ath79 and other DTS-based devices, the sysfs file is not available (as the
> > original code that exposed it was apparently deemed unsuitable for inclusion or
> > upstreaming), so these devices resort to a userspace utility ("rbextract") that
> > fetches the data from the flash and uncompresses/decodes it.
> 
> That "original code that exposed it was apparently deemed unsuitable for inclusion or upstreaming)" is problematic. Since OpenWrt wants to upstream. But I gather there was a reason why it the existing userspace utility ("rbextract") from ath79 wasn't suitable for ipq40xx.

Just to clarify as it seems you misunderstood the cover letter: the "original code" I'm referring to here is code that was written by Gabor Juhos for ar71xx, see [here](https://github.com/openwrt/openwrt/blob/9f2a40c72fdb0b9b077bf5b4b01aa17eac4bfb45/target/linux/ar71xx/files/arch/mips/ath79/routerboot.c). _That_ code was unsuitable. Not the code I wrote and got merged in OpenWRT.

The "rbextract" utility was a quick hack put together by robimarko and taken out of his tree without his approval, and was not intended to be merged and definitely not production-ready code. Hence its removal.

Which is why I don't understand your comment nor the point you're trying to make by quoting it in your last reply.

It bears repeating this PR provides a clean, working and tested solution that has no side effects to other targets and works exactly as the underlying code intends.

Yet it looks as if you're implying this PR should be NACK'd without explicitly saying so. Would you mind clarifying?

Thanks

notr1ch commented:
> @alaraun
> 
> see [#4679 (comment)](https://github.com/openwrt/openwrt/pull/4679#issuecomment-943362125) and [#4679 (comment)](https://github.com/openwrt/openwrt/pull/4679#issuecomment-942690074)

I think the discussion of the sysfs driver is out of scope of this PR. The sysfs driver is already merged and shipped in multiple OpenWrt releases, I don't think this PR should be blocked because of it.

andixlm commented:
Built a binary for MikroTik hap ac2 a while ago, maybe 2 months have passed already. No issues with wireless, performance is somewhat comparable to RouterOS, there's a slight degradation of uploading speed in the same environment (~400mbps -> ~230mbps) but, you know, it's still fast and stable enough to complain about it thus I'm ok with it

hauke commented on package/kernel/ath10k-ct/patches/100-ath10k-support-bus-and-device-specific-API-1-BDF-sel.patch:
```diff
@@ -0,0 +1,87 @@
+-		else
++			strlcpy(boardname, ar->fwcfg.bname, sizeof(boardname));
++		}
++		else {
```
The `}` and the `else {` should be in the same line

hauke reviewed (commented):
@chunkeey  How do we want to go forward with this pull request?

chunkeey commented:
> @chunkeey How do we want to go forward with this pull request?

This reads like you have something in mind?

@f00b4r0 said:
https://github.com/openwrt/openwrt/pull/4679#issuecomment-943278709
> tl;dr: @robimarko's proposal leverages the existing code as it was intended when I wrote it, and is clean as is. I think we should wait to hear back from upstream about his changes to the ath10k driver before further scratching this itch."

I haven't heard any news about this.

As for BDF. There are options like making a dynamic board-2.bin on the fly https://github.com/openwrt/openwrt/pull/4679#issuecomment-942690074 . But this idea seems dead, since nobody bothered to even try it.

So, if they want to stick to the upstream rules: Go the upstream board file route:
https://wireless.wiki.kernel.org/en/users/drivers/ath10k/boardfiles
I guess the idea here would be that a device can share a common .dtsi with just a custom qcom,ath10k-calibration-variant property in the device-variant.dts. Then all the individual boardfiles with a unique variant (i.e: -v1/-v2/ or rev1/2, etc) string gets placed into the board-2.bin. 

Note: Kalle's ath10k-firmware [board-2.bin](https://github.com/kvalo/ath10k-firmware/blob/master/QCA4019/hw1.0/board-2.bin) for the ipq40xx is now over 1 MiB+. It's getting close to double the firmware size. However these board-2.bin compress really well. I.e.: lzma'd with -6 produces a 12563 byte board-2.bin.lzma file.

This does have the benefit, that no changes to ath10k(-ct) nor the existing extractors are necessary.
A possible, but very obvious downside would be that unless Mikrotik has something like a revision labeled on the device sticker that could tell - without disassembling - whenever this is a v1,v2 (etc.), people would need to flash a image first to check if it's the right one. (this would made the fd717f54beaaa791123cd146af4e790cc19a05c6 unnecessary though.)

chunkeey commented:
I see, this reply triggered another inquiry on ath10k-devel/linux-wireless. Good :)

robimarko commented:
> > @chunkeey How do we want to go forward with this pull request?
> 
> This reads like you have something in mind?
> 
> @f00b4r0 said: [#4679 (comment)](https://github.com/openwrt/openwrt/pull/4679#issuecomment-943278709)
> 
> > tl;dr: @robimarko's proposal leverages the existing code as it was intended when I wrote it, and is clean as is. I think we should wait to hear back from upstream about his changes to the ath10k driver before further scratching this itch."
> 
> I haven't heard any news about this.

Well, it's just sitting upstream and waiting, not much more than I ping that I can do.

> 
> As for BDF. There are options like making a dynamic board-2.bin on the fly [#4679 (comment)](https://github.com/openwrt/openwrt/pull/4679#issuecomment-942690074) . But this idea seems dead, since nobody bothered to even try it.

Well, I haven't tried it as I don't find it a practical solution compared to just serving a plain BDF.
> 
> So, if they want to stick to the upstream rules: Go the upstream board file route: https://wireless.wiki.kernel.org/en/users/drivers/ath10k/boardfiles I guess the idea here would be that a device can share a common .dtsi with just a custom qcom,ath10k-calibration-variant property in the device-variant.dts. Then all the individual boardfiles with a unique variant (i.e: -v1/-v2/ or rev1/2, etc) string gets placed into the board-2.bin.

I think that @f00b4r0 and I have explained before, this can't and won't work.
The issue is that BDF-s changed but the revision hasn't, and that is just for hAP ac2, they can change others whenever they want
as its meant to be dynamic, that is why its shipped as part of the factory data and not as others do and just ship a blob in userspace.
> 
> Note: Kalle's ath10k-firmware [board-2.bin](https://github.com/kvalo/ath10k-firmware/blob/master/QCA4019/hw1.0/board-2.bin) for the ipq40xx is now over 1 MiB+. It's getting close to double the firmware size. However these board-2.bin compress really well. I.e.: lzma'd with -6 produces a 12563 byte board-2.bin.lzma file.

The size is not the primary issue, though 1MiB is kind of insane.
> 
> This does have the benefit, that no changes to ath10k(-ct) nor the existing extractors are necessary. A possible, but very obvious downside would be that unless Mikrotik has something like a revision labeled on the device sticker that could tell - without disassembling - whenever this is a v1,v2 (etc.), people would need to flash a image first to check if it's the right one. (this would made the [fd717f5](https://github.com/openwrt/openwrt/commit/fd717f54beaaa791123cd146af4e790cc19a05c6) unnecessary though.)

As I explained above, this won't work as there is no marking difference on the boards that have newer/changed BDF-s and those that have the original ones, revision stayed the same.
So there is no way to know and let alone account for all the possible variants, and I don't see a point in doing so if we can just dynamically serve it and don't worry at all.
So, I dont see what is the point of contention in this PR?

wwortel commented:
My support for Robimarko' point of view in this.
OpenWrt should let the quality of the solution prevail.
The logic:
1) the vendor does what the vendor wants,
2) the vendor's product contains the best possible solution at the time of production,
3) the vendor can introduce improvements during production and has no obligation to make those sticker identifiable,
4) clever minds have figured out how to retrieve the essential data from the vendor's product,
Dynamic loading of those data, considered best by the vendor, can be implemented and i.m.h.o. should prevail.
A file that gets loaded that contains lots of information that is irrelevant for the product at hand is at best a temporary fix and is never up to date.

The solution proposed here works flawlessly; so introducing it would be my recommendation.

neg2led commented:
This seems like a completely reasonable solution to this problem, and I can think of a couple of other places where it would be quite helpful, which I'll get to further down.

> > As for BDF. There are options like making a dynamic board-2.bin on the fly [#4679 (comment)](https://github.com/openwrt/openwrt/pull/4679#issuecomment-942690074) . But this idea seems dead, since nobody bothered to even try it.
> 
> Well, I haven't tried it as I don't find it a practical solution compared to just serving a plain BDF.

Generating board-2.bin dynamically means taking data that is already in a compressed/encoded form in flash, decoding and decompressing it, then storing it on the flash again, no? This seems like a rather silly approach; would it not ultimately be doing the same thing as this PR does, but wasting flash space and erase cycles along the way?

There may be some CPU cost when the driver loads this data, but AFAIK that only happens on boot / driver initialization, which is not exactly a frequent occurrence.

> > So, if they want to stick to the upstream rules: Go the upstream board file route: https://wireless.wiki.kernel.org/en/users/drivers/ath10k/boardfiles I guess the idea here would be that a device can share a common .dtsi with just a custom qcom,ath10k-calibration-variant property in the device-variant.dts. Then all the individual boardfiles with a unique variant (i.e: -v1/-v2/ or rev1/2, etc) string gets placed into the board-2.bin.

This is impossible, as @robimarko and @f00b4r0 have previously explained; these board revisions have completely different BDFs, but present the same exact unique identifiers to the driver; there is fundamentally no way to determine the correct BDF by querying the device.

For that matter, many ath10k devices have different board data files present in their OEM firmware for different regions or radio configurations - again with no way to programmatically determine which file to use via information available to the driver. 

A perfect example of this is the Sophos APX120 and APX320/530/740 series (ipq4019 and ipq8064 respectively), which are the devices I mentioned at the top of this comment. All four of these models have ath10k radios, and come with no fewer than 7 different board data files for each radio, all sharing the same identifiers. 

The different files are primarily used to adjust calibration for different regions (FCC, ETSI, IL, IC, CA, AU, BR, CR - AU is identical to ETSI but gets its own folder for some reason) but they serve another purpose; on the APX320/530/740 the 2.4GHz radio can optionally run as a second 5GHz radio, and it uses a unique board data file (actually another one per region) to do this. A unique file which, again, has the same board ID and an indistinguishable set of identifiers to the others.

That's not _quite_ the same problem this PR is attempting to solve - in the case of the Sophos devices, the board data variants are not all available in the device's flash, only the device's original factory region files are - but it does make it impossible to upstream the board data files. There is no way I'm getting somewhere north of 60 unique board data files *that cannot be differentiated from each other through board/chipset identifiers alone* upstreamed, let alone just to support four devices.

> there is no way to know and let alone account for all the possible variants, and I don't see a point in doing so if we can just dynamically serve it and don't worry at all. So, I dont see what is the point of contention in this PR?

Agreed, I am not sure what the argument is here. This seems to be a reasonable solution that's in-line with kernel interface/behaviour guidelines for firmware loading fallbacks; it may even be upstreamable if the ML discussion is anything to go by, but if anything that seems like an argument in favor of merging it in and dropping it if/when upstream applies it?

chunkeey commented:
Again, you guys. OpenWrt is not upstream. Take your concerns and opinions up with the folks over at [linux-wireless@vger.kernel.org](https://wireless.wiki.kernel.org/en/developers/mailinglists). This is the right way to do it!

That said, I'm happy that people are waking up and realize how "great" these whole pre-cal + board stuff is.
Back in 2016, I too had that with the RT-AC58U. Upstream solution was the VARIANT property for all of this.
Yes, It solved the RT-AC58U problem (and many more devices that came after)... in a way...

As for @f00b4r0 caldata_from_sysfs. I can tell you why I am/was so confused about this: "The same sysfs 
mechanism" is used by procd's hotplug already in handle_firmware(). See:
<https://git.openwrt.org/?p=project/procd.git;a=blob;f=plug/hotplug.c;h=beff38aaaef2b969e968742ee449b893c9405e84;hb=HEAD#l224>

The way this is supposed to work is that
  1. some driver does a request_firmware("chip-firmware.bin"))
  2. kernel looks if firmware file is present in /lib/firmware/chip-firmware.bin. 
       If the file exists, the kernel reads the file itself and provides it to the driver (kernel can decompress firmware files if enabled)
  3. if the chip-firmware.bin doesn't exist, the firmware_loader will invoke the fallback mechanism and notify userspace through uevent
  4. openwrt's procd will get the uevent, look up in /etc/hotplug-preinit.json (during early boot) and later /etc/hotplug.json
      what to do for the "FIRMWARE" action . [...]

                                [
                                        [ "exec", "/sbin/hotplug-call", "%SUBSYSTEM%" ],
                                        [ "load-firmware", "/lib/firmware" ],
                                        [ "return" ]
                                ]

  5. hotplug-call firmware -> this executes the extractor scripts in /etc/hotplug.d/firmware/*
      (for ipq40xx, this is just our beloved /etc/hotplug.d/firmware/11-ath10k-caldata file)

  6. 11-ath10k-caldata then does what ever the device needs. The caldata_extract* scripts then write
       the requested "chip-firmware.bin" as a file to /lib/firmware/chip-firmware.bin.
       (On the next boot the kernel will basically be able to.)
 
      Yes, that file can be anything under /lib/firmware.  Currently ipq40xx uses this only for (pre-)caldata.
      But ath79 devices like comfast,cf-wr752ac-v1 already delete their board-2.bin and overwrite with pre-cal as
      board.bin for some qca9888 that also suffer this.

      This is why I'm so adamant about the https://github.com/openwrt/openwrt/pull/4679#issuecomment-942690074
      It is possible to screw with the board-2.bin on the fly!

  7. the "load-firmware" instructs procd to call that handle_firmware() which will then see that there's the file in
      /lib/firmware/... and upload it through the sysfs firmware machanism.

The caldata_from_sysfs() does what procd would be doing in step 7. Except, it doesn't store a /lib/firmware/chip-firmware.bin
file. On every boot step 1-7 have to be done and at the end procd hotplug throws "Could not find firmware chip-firmware.bin" error for good measure. "Is this fine?"

chunkeey commented:
Quick update: The discussion does go on.

In the meantime, I made a updated `/etc/hotplug.d/firmware/11-ath10k-caldata` that does the magic as well.
This doesn't need any backports or kernel patches. Yes: It works with existing installations. You just need
access to the device and a way to either download a file from git.openwrt.org (https) or copy it to the device
and run commands there.

For interested parties:

1. download [11-ath10k-caldata](https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=blob_plain;f=target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata;h=57c8d7a610ced7b4dc195ed14708ee58dbfa467c;hb=52f3407d94da62b99ba6c09f3663464cccd29b4f) replacement.

This replaces the file with the same `11-ath10k-caldata` name on your device in `/etc/hotplug.d/firmware/`
One way to do it on the device would be to install curl (it automatically pulls in ca-bundle + ssl). 

    opkg update; opkg install curl 
    curl "https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=blob_plain;h=57c8d7a610ced7b4;hb=52f3407d94da62b" > /etc/hotplug.d/firmware/11-ath10k-caldata

2. delete the existing /lib/firmware/ath10k/QCA4019/hw1.0/board-2.bin
(this is necessary, so the board-2.bin can be created by the next ath10k_pci invocation)

    rm /lib/firmware/ath10k/QCA4019/hw1.0/board-2.bin

3. reload ath10k_pci (alternatively: reboot)

    rmmod ath10k_pci; modprobe ath10k_pci

Please report back your findings. This was tested on a ipq40xx Mikrotik device can serve as an intermediate method until... well... no idea.

f00b4r0 commented:
> Please report back your findings. This was tested on a ipq40xx Mikrotik device can serve as an intermediate method 

Not sure how *this* intermediate method is any better than the PR's. It will bloat the flash and will interfere if/when this PR is eventually merged.

> until... well... no idea.

Would you mind clearly stating your intent? Do we have to read between these lines that you are implicitly NACK'ing this PR? Would you mind elaborating on the *technical* reasons to do so, when it incidentally fixes a driver bug and is proposed by the very people who actually care about these devices?

Thanks

markmentovai commented:
@chunkeey https://github.com/openwrt/openwrt/pull/4679#issuecomment-1042401764:
> 1. download [11-ath10k-caldata](https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=blob_plain;f=target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata;h=57c8d7a610ced7b4dc195ed14708ee58dbfa467c;hb=52f3407d94da62b99ba6c09f3663464cccd29b4f) replacement.

This seems somewhat inelegant given all of the appending going on. What happens if two `11-ath10k-caldata` instances run simultaneously? They race, and in your words

@chunkeey https://github.com/openwrt/openwrt/pull/4679#issuecomment-1029926837:
> It is possible to screw with the board-2.bin on the fly!

Exactly.

Having reviewed both, I prefer @robimarko’s solution in this pull request (#4679). Your proposal and his are nearly identical in all respects except yours takes an extra step to merge two radios’ BDFs into a single `board-2.bin`. By keeping them independent, his has a robustness edge.

chunkeey commented:
> @chunkeey [#4679 (comment)](https://github.com/openwrt/openwrt/pull/4679#issuecomment-1042401764):
> 
> > 1. download [11-ath10k-caldata](https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=blob_plain;f=target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata;h=57c8d7a610ced7b4dc195ed14708ee58dbfa467c;hb=52f3407d94da62b99ba6c09f3663464cccd29b4f) replacement.
> 
> This seems somewhat inelegant given all of the appending going on. What happens if two `11-ath10k-caldata` instances run simultaneously? They race, and in your words
> 

Look in your dmesg/bootlog of your favorite ipq40xx (this case is a NBG6617, though this is likely not too different from any other): 

```
[   15.846421] ath10k_ahb a000000.wifi: qca4019 hw1.0 target 0x01000000 chip_id 0x003b00ff sub 0000:0000
[   15.846477] ath10k_ahb a000000.wifi: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   15.858356] ath10k_ahb a000000.wifi: firmware ver 10.4b-ct-4019-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 6b2b5c5b
[   15.953688] ath10k_ahb a000000.wifi: board_file api 2 bmi_id 0:16 crc32 e2dfaa91
[   17.241418] ath10k_ahb a000000.wifi: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
[   17.241470] ath10k_ahb a000000.wifi: msdu-desc: 2500  skid: 32
[   17.288594] ath10k_ahb a000000.wifi: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
[   17.289582] ath10k_ahb a000000.wifi: wmi print 'free: 53252 iram: 13432 sram: 35752'
[   17.446334] ath10k_ahb a000000.wifi: htt-ver 2.2 wmi-op 6 htt-op 4 cal pre-cal-file max-sta 32 raw 0 hwcrypto 1
[   17.626995] ath: EEPROM regdomain sanitized
[   17.627013] ath: EEPROM regdomain: 0x64
[   17.627022] ath: EEPROM indicates we should expect a direct regpair map
[   17.627035] ath: Country alpha2 being used: 00
[   17.627041] ath: Regpair used: 0x64
...1.5 second pause...
[   19.021944] ath10k_ahb a800000.wifi: qca4019 hw1.0 target 0x01000000 chip_id 0x003b00ff sub 0000:0000
[   19.022012] ath10k_ahb a800000.wifi: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   19.035348] ath10k_ahb a800000.wifi: firmware ver 10.4b-ct-4019-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 6b2b5c5b
[   19.089633] ath10k_ahb a800000.wifi: board_file api 2 bmi_id 0:17 crc32 e2dfaa91
[   20.378474] ath10k_ahb a800000.wifi: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
[   20.378536] ath10k_ahb a800000.wifi: msdu-desc: 2500  skid: 32
[   20.425895] ath10k_ahb a800000.wifi: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
[   20.427543] ath10k_ahb a800000.wifi: wmi print 'free: 53252 iram: 13432 sram: 35752'
[   20.565345] ath10k_ahb a800000.wifi: htt-ver 2.2 wmi-op 6 htt-op 4 cal pre-cal-file max-sta 32 raw 0 hwcrypto 1
[   20.724700] ath: EEPROM regdomain sanitized
[   20.724724] ath: EEPROM regdomain: 0x64
[   20.724733] ath: EEPROM indicates we should expect a direct regpair map
[   20.724752] ath: Country alpha2 being used: 00
[   20.724761] ath: Regpair used: 0x64
[   20.732047] kmodloader: done loading kernel modules from /etc/modules.d/*
```

Yes: a800000 starts probing after a000000 finished. The race gets called off before it begins.
(this is partly because of a special openwrt ath10k patch and because of procd firmware_handler).

> @chunkeey [#4679 (comment)](https://github.com/openwrt/openwrt/pull/4679#issuecomment-1029926837):
> 
> > [...] screw [...] board[..].bin!
> 
> Exactly.

Exactly!

> 
> Having reviewed both, I prefer @robimarko’s solution in this pull request (#4679). Your proposal and his are nearly identical in all respects except yours takes an extra step to merge two radios’ BDFs into a single `board-2.bin`. By keeping them independent, his has a robustness edge.

No, sorry. I don't mind being taken out of or in context. 

Again, my take is that you either feed upstream their board.bin's.... or not to play the upstream game and come up with ingenious solutions for a problem that shouldn't exist in the first place.

robimarko commented:
> 
> Again, my take is that you either feed upstream their board.bin's.... or not to play the upstream game and come up with ingenious solutions for a problem that shouldn't exist in the first place.

What do you mean by this?
If you mean pushing the BDF-s for MikroTik devices upstream, then I think that it's been made really clear that it's not possible.
How are users gonna know that they need a different BDF?

Boards aren't marked any different, heck even the revision is the same.
I honestly don't understand what is the issue with dynamically loading the BDF-s?

alaraun commented:
@hauke I would like to get this one approved and merged. At this point it is the best option to go with Mikrotik devices.

notr1ch commented:
Is there anything I can do to help move this along? What is the current hold up?

I'll stress again that this is absolutely *required* on modern Mikrotik Wi-Fi devices. I'd suggest that Wi-Fi should be listed as unsupported on the affected device pages otherwise, as it's completely unusable without this patch.

robimarko commented:
There was finally movement on the ath10k patch upstream, once that gets merged its just a matter of backporting it.

robimarko commented:
ath10k patch was merged upstream, PR has been rebased and should be ready to go.

f00b4r0 commented:
@robimarko I don't think you need to update all the hunks in unrelated patches: it needlessly clutters your commit and make it look as if it touches other parts which it really doesn't. Patch will be able to handle the offsets on its own. My 2c :)

f00b4r0 commented:
I don't think anybody will. Especially for a backport patch that will eventually go away. But that's your call :)

robimarko commented:
I had cases like those in general before, so now I automatically refresh patches.

hauke commented on target/linux/ipq40xx/Makefile:
```diff
@@ -19,6 +19,6 @@ DEFAULT_PACKAGES += \
 	kmod-ath10k-ct wpad-basic-wolfssl \
 	kmod-usb3 kmod-usb-dwc3 ath10k-firmware-qca4019-ct \
-	ath10k-board-qca4019 uboot-envtools
+	uboot-envtools
```
This will also remove it for the `chromium` subtarget. I think it is needed there.

hauke replied:
I think you can also add `-ath10k-board-qca4019` to the mikrotik subtarget.

robimarko replied:
Indeed, this PR predated the Chromium target, let me check whether removing the package in mikrotik subtarget works as I think I did not work when this was initially developed.

Also, need to update the ath10k-ct as Ben merged the backport.

robimarko replied:
Done, removing it in the subtarget appears to work(Probably forgot to purge tmp before).
ath10k-ct is updated instead of adding a patch.

It should be ready now.

john-tho replied:
> I think you can also add `-ath10k-board-qca4019` to the mikrotik subtarget.

I am not sure if this will work with BUILDBOT config?
In a quick grep, I could not find any examples of packages being `-cut` from `DEFAULT_PACKAGES`, and I have run into issues with `DEVICE_PACKAGES` where `-cut` packages are ignored by BUILTBOT config for the initramfs images.

robimarko replied:
Initramfs images with `CONFIG_TARGET_MULTI_PROFILE` and `CONFIG_TARGET_PER_DEVICE_ROOTFS` don't consider per device packages AFAIK, its a buildroot limitation.

I just tried adding `CONFIG_TARGET_MULTI_PROFILE` and `CONFIG_TARGET_PER_DEVICE_ROOTFS` and building for the MikroTik subtarget and it did not include the upstream board-2.bin in the squashfs

john-tho replied:
Confirming that the buildbot snapshot initram image with these changes (r19677 5eee67a72f) does include, and use `ath10k-board-qca4019`
```
[    9.682562] ath10k_ahb a000000.wifi: board_file api 2 bmi_id 0:16 crc32 11892f9b
[   12.258448] ath10k_ahb a800000.wifi: board_file api 2 bmi_id 0:17 crc32 11892f9b
```
Would you, or @hauke be against a patch to remove `ath10k-board-qca4019` from the target makefile, and add it to the generic and chromium subtargets? I do testing from initram images (especially on unreachable devices), and would appreciate the more consistent wireless performance between the initram and installed images.

Thank you for your work on these changes!

robimarko replied:
Thanks for checking, no idea why it worked as now with `CONFIG_BUILDBOT` it ignores the subtarget device package removal.
I don't have anything against that, it's probably the only way to do this properly.
If you chose to make a patch, can you send it to the mailing list, that is usually faster to get merged

robimarko replied:
Otherwise I will make a patch since this is a mess that I created

robimarko committed:
ath10k-ct: update to 2022-05-13

Update ath10k-ct to the latest version which includes the backported
ath10k commit for requesting API 1 BDF-s with a unique name like caldata.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
ipq40xx: mikrotik: dont include ath10k-board-qca4019 by default

Since MikroTik subtarget now uses dynamic BDF loading its crucial that it
doesnt include the board-2.bin at all which is provided by the
ath10k-board-qca4019 package.

So to resolve this dont include the ath10k-board-qca4019 package on the
MikroTik subtarget.

Signed-off-by: Robert Marko <robimarko@gmail.com>

hauke commented:
Thank you for the patches, I applied them to master in ab97b2a25d69215dcc0d9621e491aa7b17f315cb, 3daf2d477ee728b5b066fe7f31808a5f19bb98a1, 4d4462cc2ace4b044e99e9b22a24cb4d89c7db95, ab141a6e2cb645ff64adb107af2e8973a720c1c7 and 5eee67a72fed52ac686dd467d93eea95d44c8dff.
